technology-considerations/API_management_AppSync_GraphQL_REST.md

```diff
@@ -0,0 +1,62 @@
+# Options for API management
+
+We want to have an intermediate component between our frontend (which consists of the browser and the Play web application)
+and the database. 
+This component will represent our APIs. The APIs will include virus scanning, format checking, validations of the format.
+
+The APIs can use the [REST architectural style]  (https://en.wikipedia.org/wiki/Representational_state_transfer),
+or we could also take advantage of [GraphQL](https://graphql.org/), that introduces a schema and a type system, along with greater flexibility.
+
+We decided to explore using GraphQL for the following reasons:
+- we anticipate multiple datasources, a relational database for keeping the information about files, transfers, users 
+ and maybe a nosql solution for auditing and GraphQL will make it very easy to connect to them
+- we anticipate that it will reduce the amount of data we retrieve from the database, and we won't have to do
+multiple calls to get department, series, files for example
+- GraphQL is [strong typed](https://graphql.org/learn/schema/). We get a schema that validates the type of the parameters of the query
+- comes with nice feature that will speed up development, like out of the box support for [pagination](https://graphql.org/learn/pagination/)
+
+For using GraphQL, we have two different options:
+
+Option 1. Use AppSync, which is a service managed by AWS for a graphQL server
+
+![appsync](images/AppSync.png "Frontend connects to AWS Appsync")
+
+Advantages:
+- it's managed by Amazon, so we don't have to worry about a server
+- it integrates with AWS Cognito, so we can use the authorisation process that Cognito provides
+- websockets out of the box, which is a mechanism to send real time updates to the user
+Disadvantages:
+- Big lockin to an AWS solution
```

Comment:
Some related disadvantages are that [AppSync was launched less than two years ago](https://aws.amazon.com/blogs/aws/introducing-amazon-appsync/) so the tools around it might be immature, and it's long-term support is less certain than for the other options.

Reply:
Added